### PR TITLE
[Sysrst_ctrl] Added EC and power-on-reset testcase

### PR DESCRIPTION
**Pull request by patel-madhuri**
This test will verify the EC and power-on-reset feature.

Signed-off-by: Madhuri Patel <madhuri.patel@ensilica.com>

### CODE REVIEW

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// This sequence will verify the ec and power on reset sequence.
+class sysrst_ctrl_ec_pwr_on_rst_vseq extends sysrst_ctrl_base_vseq;
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:6]: this seq is really similar to the wr_prot seq. can you try to make a sequence or task for them, to avoid the duplication.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+   task body();
+
+     bit allowed_ec_rst_l_1, en_override_ec_rst_l;
+
+    `uvm_info(`gfn, "Starting the body from ec_pwr_on_rst_vseq", UVM_LOW)
+
+     // Ensure EC is in reset when OT is out of reset
+     csr_rd(ral.pin_allowed_ctl.ec_rst_l_1, allowed_ec_rst_l_1);
+     csr_rd(ral.pin_out_ctl.ec_rst_l, en_override_ec_rst_l);
+     if (allowed_ec_rst_l_1 == 0 & en_override_ec_rst_l == 1) begin
```
Comment (matutem) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:22]: If this condition doesn't hold there is no check of ec_rst_l_out. Should this condition always hold after reset? If so, change it to `DV_CHECKs. Is there any randomization possible?

File: hw/ip/sysrst_ctrl/dv/tb.sv [hunk 3]
```diff
@@ -82,4 +82,18 @@ module tb;
     run_test();
   end
 
+  initial begin
```
Comment (matutem) [inline hw/ip/sysrst_ctrl/dv/tb.sv:85]: It seems these assertions should go under dv/sva.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      `DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 0);
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:23]: same as this [comment](https://github.com/lowRISC/opentitan/pull/9928#discussion_r791211434)

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+     // Explicitly release the EC reset
+     // Disable the override function
+     ral.pin_out_ctl.ec_rst_l.set(0);
+     csr_update(ral.pin_out_ctl);
+
+     // Configure the EC_RST_CTL register to stretch the EC rst
+     csr_wr(ral.ec_rst_ctl, set_ec_rst_timer);
+
+     wait(cfg.vif.ec_rst_l_out == 1'b1);
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:35]: we should wait for specific aon clock rather than the output pin, so that we know the timer works well. For example, if we set timer to 10 cycles, but it takes 20 cycles, then it's not right. But we won't catch it if we simply wait until it happens.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+     cfg.clk_aon_rst_vif.wait_clks(1);
+    `DV_CHECK_EQ(cfg.vif.ec_rst_l_in, cfg.vif.ec_rst_l_out);
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:38]: ideally, it's to randomize all inputs that affect this output pin ec_rst_l_in, then check it with the expected value.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv (hunk 2 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:22]: This condition should always be true after reset as these are the reset values of the register fields. I will add `DV_CHECKs

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv (hunk 5 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:35]: I will add cfg.clk_aon_rst_vif.wait_clks(set_ec_rst_timer);

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv (hunk 6 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:38]: To explicitly release the reset we are disabling the override function and configuring the ec_rst_ctl register. The ec_rst_l_o is expected to deassert once the timer has finished. May be @msfschaffner could comment in more details.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv (hunk 2 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:22]: This check is already done in csr_hw_reset test so I am removing this from this test

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// This sequence will verify the ec and power on reset sequence.
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:5]: can you explain a bit more detail about the sequence and checks

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      // Ensure EC is in reset when OT is out of reset
+      `DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 0);
+
+      // Explicitly release the EC reset
+      // Disable the override function
+      ral.pin_out_ctl.ec_rst_l.set(0);
+      csr_update(ral.pin_out_ctl);
+
+      // Configure the EC_RST_CTL register to stretch the EC rst
+      csr_wr(ral.ec_rst_ctl, set_ec_rst_timer);
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:28]: > ec_rst_pulse | Configure the pulse width of ec_rst_l. 10-200ms, each step is 5us(200KHz clock)

From the spec, this CSR will configure the pulse width of `ec_rst_l`. The combo detect may generate this pulse, but I don't find input `ec_rst_l` can also trigger this pulse.

@msfschaffner can you also take a look at this sequence?

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+   rand uint16_t set_ec_rst_timer;
+
+   task body();
+
+    `uvm_info(`gfn, "Starting the body from ec_pwr_on_rst_vseq", UVM_LOW)
+
+    repeat ($urandom_range(1,20)) begin
+
+      // Ensure EC is in reset when OT is out of reset
+      `DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 0);
```
[the remark this answers is not in the document]
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:20]: Yes, I can move  "`DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 0);"  check out of loop. The reset value of cfg.vif.ec_rst_l_in is 1 so once the timer is finished we are expecting the value of cfg.vif.ec_rst_l_o to be 1 from reset value 0. Thus we check  "`DV_CHECK_EQ(cfg.vif.ec_rst_l_in, cfg.vif.ec_rst_l_out);" after waiting for the timer.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv (hunk 8 above)
Comment (msfschaffner) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:34]: This is a hardwired functionality that is also implemented in the combo module (although it is not a combination).
Whenever `cio_ec_rst_l_i` transitions from 1 -> 0, this action is automatically triggered.

The use case here is that `sysrst_ctrl` monitors this system-level reset signal, and whenever it goes low, we stretch it to the minimum pulse width configured, so that we can ensure the reset propagates properly to the reset of the system.

Comment (msfschaffner) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:34]: I will add some more description regarding hardwired functionality to the spec.

Comment (weicaiyang) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:34]: Thanks @msfschaffner for clarifying this in the spec

Looks like we should drive ec_rst_l from 1->0 and check ec_rst_l_out is high for a mount of cycles as configured in CSR `EC_RST_CTL`

Comment (msfschaffner) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:34]: @weicaiyang the ec_rst_l signal is an active low signal (the naming convention does not follow the guidelines properly here and uses l instead of n).

Hence, I believe we see the following in the simulation:
1) the device comes out of reset, and ec_rst_l is asserted (low) automatically
2) the testbench disables the ec_rst_l assertion
3) the reset stretching circuit continues to keep ec_rst_l low until it has expired (even though the software has discabled the ec_rst_l assertion)
4) once the reset stretching is complete, ec_rst_l is deasserted (high)
5) a few cycles afterwards, a transition from high to low is detected on the ec_rst_l input, causing the reset stretching circuitry to again assert the reset signal low for many cycles

So basically what we are seeing here appears to be two stretched resets in a row.

Comment (weicaiyang) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:34]: It makes sense! Thanks for the explanation. I finally understand how this works.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      ral.pin_out_ctl.ec_rst_l.set(0);
+      csr_update(ral.pin_out_ctl);
+
+      // Configure the EC_RST_CTL register to stretch the EC rst
+      csr_wr(ral.ec_rst_ctl, set_ec_rst_timer);
+
+      cfg.clk_aon_rst_vif.wait_clks(set_ec_rst_timer);
+
+      cfg.clk_aon_rst_vif.wait_clks(1);
+      `DV_CHECK_EQ(cfg.vif.ec_rst_l_in, cfg.vif.ec_rst_l_out);
```
[the remark this answers is not in the document]
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:39]: The reset value of ec_rst_l_in is 1 and ec_rst_l_out stays low till the configured timer value. Once the timer is finished, value of ec_rst_l_out should be 1. Thus I am checking at line no 39 if ec_rst_l_out is 1.
I should actually change it to **`DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 1);**

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+// This sequence will verify the ec and power on reset sequence.
+// It will explicitly release the reset by disabling the override
+// function and stretching the ec reset to the value configured in
+// ec_rst_ctl register
+class sysrst_ctrl_ec_pwr_on_rst_vseq extends sysrst_ctrl_base_vseq;
+  `uvm_object_utils(sysrst_ctrl_ec_pwr_on_rst_vseq)
+
+  `uvm_object_new
+
+   rand uint16_t set_ec_rst_timer;
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:14]: let's constrain timer to small value for most of times. 
```
constraint set_ec_rst_timer_c {
    set_ec_rst_timer dist {
      [10:100] :/ 95,
      [101:$]   :/ 5
    };
}
```

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    repeat ($urandom_range(2,4)) begin
+
+      // Explicitly release the EC reset
+      // Disable the override function
+      ral.pin_out_ctl.ec_rst_l.set(0);
+      csr_update(ral.pin_out_ctl);
+
+      // Configure the EC_RST_CTL register to stretch the EC rst
+      csr_wr(ral.ec_rst_ctl, set_ec_rst_timer);
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:28]: we can move the CSR program out of the loop, since it uses the same value.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      // Check ec_rst_l_out is not driven high
+      `DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 0);
+
+      // Drive ec_rst_l_in high again
+      cfg.vif.ec_rst_l_in = 1;
+    end
+    
+    cfg.clk_aon_rst_vif.wait_clks(20);
+   endtask : body
+
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:49]: Let's create a task to monitor ec_rst_l asserts low for expected cycles.
```
task monitor_ec_rst_low(int exp_cycles);
  int act_cycles;
  int aon_period_ns = cfg.clk_aon_rst_vif.clk_period_ps / 1000;
  // wait for ec_rst_l_out===0
  `DV_SPINWAIT(wait(cfg.vif.ec_rst_l_out === 0), , aon_period_ns * 10)
  // check ec_rst_l_out is low for exp_cycles. After exp_cycles+10, below will time out and fail.
  `DV_SPINWAIT(while (cfg.vif.ec_rst_l_out === 0) act_cycles++;, , aon_period_ns * exp_cycles + 10)
endtask
```
split the drive 1->0 for ec_rst_l_in to a task, as it should be in a separate thread. The timer starts to counter after falling edge of ec_rst_l_in
```
  task driver_ec_rst_l_in(int cycles);
    cfg.vif.ec_rst_l_in = 0;

    // a few extra falling edges won't affect anything
    if ($urandom_range(0, 2)) begin
      cfg.clk_aon_rst_vif.wait_clks(1);
      cfg.vif.ec_rst_l_in = 1;
      cfg.clk_aon_rst_vif.wait_clks(1);
      cfg.vif.ec_rst_l_in = 0;
    end
    cfg.clk_aon_rst_vif.wait_clks(cycles);
  endtask
```
```
   task body();

    `uvm_info(`gfn, "Starting the body from ec_pwr_on_rst_vseq", UVM_LOW)


    // Explicitly release the EC reset
    // Disable the override function
    ral.pin_out_ctl.ec_rst_l.set(0);
    csr_update(ral.pin_out_ctl);

    // Configure the EC_RST_CTL register to stretch the EC rst
    csr_wr(ral.ec_rst_ctl, ec_rst_timer);

    // check ec_rst_l asserts for ec_rst_timer cycles after reset
    monitor_ec_rst_low(set_ec_rst_timer);
    repeat ($urandom_range(2, 4)) begin
      cfg.clk_aon_rst_vif.wait_clks($urandom_range(1, 100));

      // remains high
      `DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 1);

      fork
        driver_ec_rst_l_in($urandom_range(1, ec_rst_timer * 2));
        monitor_ec_rst_low(ec_rst_timer);
      join
    end
```

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv [hunk 14]
```diff
@@ -19,6 +19,29 @@ class sysrst_ctrl_base_vseq extends cip_base_vseq #(
     cfg.clk_aon_rst_vif.set_freq_khz(200);
   endtask
 
+  virtual task monitor_ec_rst_low(int exp_cycles);
+    int act_cycles;
+    int aon_period_ns = cfg.clk_aon_rst_vif.clk_period_ps / 1000;
+    // check ec_rst_l_out is low for exp_cycles. After exp_cycles+10, below will time out and fail.
+    `DV_SPINWAIT(while (cfg.vif.ec_rst_l_out != 1) begin
+                 cfg.clk_aon_rst_vif.wait_clks(1);
+                 act_cycles++;
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv:28]: nit: add 2 spaces indent

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv [hunk 15]
```diff
@@ -19,6 +19,29 @@ class sysrst_ctrl_base_vseq extends cip_base_vseq #(
...
+                 end, , aon_period_ns * (exp_cycles + 3))
+
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv:30]: sorry, I forgot to add a check for `act_cycles`.
```
`DV_CHECK(act_cycles inside {[exp_cycles - 3 : exp_cycles + 3]})
```

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv [hunk 16]
```diff
@@ -52,7 +75,10 @@ class sysrst_ctrl_base_vseq extends cip_base_vseq #(
         super.apply_reset(kind);
         cfg.clk_aon_rst_vif.apply_reset(0,400,0,1);
       join
-      `uvm_info(`gfn, "Driving from apply_reset", UVM_NONE)
+     // Ensure flash wp is in reset when opentitan is out of reset
+     `DV_CHECK_EQ(cfg.vif.flash_wp_l, 0);
+     // Ensure EC is in reset when OT is out of reset
+     `DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 0);
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv:81]: looks like these have been added in #9928.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+      // remains high
+      `DV_CHECK_EQ(cfg.vif.ec_rst_l_out, 1);
+
+      fork
+       begin
+        driver_ec_rst_l_in($urandom_range(1, set_ec_rst_timer * 2));
+        monitor_ec_rst_low(set_ec_rst_timer);
+       end
+      join
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_ec_pwr_on_rst_vseq.sv:49]: nit: indent

**Review by weicaiyang — COMMENTED** [review]
Overall LGTM, add a few small comments and please resolve the merging conflict.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv [hunk 18]
```diff
@@ -19,6 +19,34 @@ class sysrst_ctrl_base_vseq extends cip_base_vseq #(
...
+    `uvm_info(`gfn, $sformatf("Value of cycles:%0d", cycles), UVM_NONE)
+    cfg.vif.ec_rst_l_in = 0;
+    // a few extra falling edges won't affect anything
+    if ($urandom_range(0, 2)) begin
+      cfg.clk_aon_rst_vif.wait_clks(1);
+      cfg.vif.ec_rst_l_in = 1;
+      cfg.clk_aon_rst_vif.wait_clks(1);
+      cfg.vif.ec_rst_l_in = 0;
+    end
+    cfg.clk_aon_rst_vif.wait_clks(cycles);
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_base_vseq.sv:47]: we should set cfg.vif.ec_rst_l_in back to 1, otherwise, ec_rst_l_out won't occur in the next loop.

**Review by weicaiyang — APPROVED** [review]
LGTM
